### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by ekinhbayar**
Includes coding style fixes along with `Client/RequestTest` and `Storage/RedisCounterTest` so far. @PeeHaa, would you please review and hint me on what else you think needs to be tested? :-)

### CODE REVIEW

File: tests/Client/RequestTest.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        $this->assertSame('Promise', get_class($promise));
+    }
+
+    public function testListenReturnsPromise()
+    {
+        $promise = yield $this->gitamp->listen();
+
+        $this->assertSame('Promise', get_class($promise));
+    }
+}
```
Comment (PeeHaa) [inline tests/Client/RequestTest.php]: New line

File: tests/Storage/RedisCounterTest.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    public function testDecrement()
+    {
+        yield $this->redis->set('test', 3);
+        yield $this->redis->decrement('test');
+        $count = yield $this->redis->get('test');
+
+        $this->assertEquals(2, $count);
+    }
+}
```
Comment (PeeHaa) [inline tests/Storage/RedisCounterTest.php]: Newline

File: tests/Client/RequestTest.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    private $gitamp;
+
+    public function setUp()
+    {
+        $this->gitamp = new GitAmp(new Client, new \ekinhbayar\GitAmp\Github\Token('token'), new Factory());
+    }
+
+    public function tearDown()
+    {
+        $this->gitamp = null;
```
Comment (PeeHaa) [inline tests/Client/RequestTest.php]: Don't think this one is needed as it's getting overwritten anyway in the setUp method calls

File: tests/Client/RequestTest.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<?php declare(strict_types=1);
+
+namespace ekinhbayar\GitAmpTests\Client;
+
+use Amp\Artax\Client;
+use ekinhbayar\GitAmp\Client\GitAmp;
+use ekinhbayar\GitAmp\Events\Factory;
+use PHPUnit\Framework\TestCase;
+
+class RequestTest extends TestCase
```
Comment (PeeHaa) [inline tests/Client/RequestTest.php]: `class GitAmpTest extends TestCase` this way it's easier to find out what class it belongs to when something fails.

File: tests/Client/RequestTest.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public function testAuthHeader()
+    {
+        $header = yield $this->gitamp->getAuthHeader();
+
+        $this->assertSame('Basic token', $header);
+    }
+
+    public function testRequestReturnsPromise()
```
Comment (PeeHaa) [inline tests/Client/RequestTest.php]: Not so much related to this test specific , but why is `GitAmp::request()` public?

It's only used inside the `GitAmp` class right? Probably want to make it `private`.

File: tests/Storage/RedisCounterTest.php [hunk 6]
```diff
@@ -4,38 +4,47 @@
...
     }
 
     public function testDecrement()
     {
-        yield $this->redis->set('test', 3);
-        yield $this->redis->decrement('test');
-        $count = yield $this->redis->get('test');
+        $this->assertEquals(1, wait($this->redis->get('test')));
+
+        $this->redis->decrement('test');
```
Comment (PeeHaa) [inline tests/Storage/RedisCounterTest.php]: Doesn't this need an `\Amp\Wait`?

File: tests/Client/RequestTest.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+        $this->assertSame('Promise', get_class($promise));
+    }
+
+    public function testListenReturnsPromise()
+    {
+        $promise = yield $this->gitamp->listen();
+
+        $this->assertSame('Promise', get_class($promise));
+    }
```
Comment (PeeHaa) [inline tests/Client/RequestTest.php]: Add some more tests to catch the other code paths:

- [ ] `testListenThrowsOnInvalidCredentials`
- [ ] `testListenReturnsResults`

**Comment by PeeHaa** [comment]
I messed up the review I think. I started a review, but accidentally click "single comment" and my review was turned into separate comments. :-(

I've also added some comments on "outdated" commits which should still be checked.

**Comment by PeeHaa** [comment]
@ekinhbayar I have fixed the GitAmp test https://gist.github.com/PeeHaa/796068f5e3b3e4c160e6dedf079dd5f2.

It now uses mocks which has several benefits:

- the unit tests never actually call the github service, but instead use mocks which makes the tests:
  - run faster
  - independent on an external service
  -  independent of secret credentials
  - test only *this* specific class and not the entire amp stack (e.g. the artax client)
- the mocks can throw exception where needed for testing purposes
- we have total control over the flow when testing
- we have isolated tests now

The same idea applies to the RedisCounter test. Instead of relying on redis being available. And running non isolated tests like which also include amp/redis we can instead use a mock RedisClient without ever going through the amp/redis stack or ever sending actually things over to a redis server.

I didn't write the redis tests yet, because I thought you may want to try to rewrite them using mocks.

File: tests/Response/ResultsTest.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    public function testAppendResponseThrowsOnInvalidJSON()
+    {
+        $response = $this->createMock(Response::class);
+
+        $response
+            ->expects($this->once())
+            ->method('getBody')
+            ->will($this->returnValue('[{"message"}]'))
+        ;
+
```
Comment (ekinhbayar) [inline tests/Response/ResultsTest.php]: This probably misses something to be actually useful and that I can't think of at this hour.

**Pull request merged**